Adding quote_style option. Fixes #293.

aendrew opened the pull request:
See #293. Also includes new test fixture. Two notes:

1. I didn't write a test for the default case (`-q 0`). Please let me know if I should do so.
1. I used the same semi-descriptive flags that `uglifyjs` does — i.e., `0`, `1`, `2`, `3`. Please let me know if I should change it to something more intuitive like `single`, `double`, `original`, `default`.

aendrew commented:
Not sure why Travis is failing, I haven't touched the failing tests. Note that the same tests failed upon cloning to local.

XhmikosR commented:
Travis failures are unrelated, I'm on it.

XhmikosR commented on Gruntfile.js:
```diff
@@ -338,6 +362,9 @@ module.exports = function(grunt) {
     'uglify:expression_js',
+    'uglify:quotes_single',
+    'uglify:quotes_double',
+    'uglify:quotes_original',
```
@vladikoff: how about we make this just `uglify` instead of specifying each subtarget?

XhmikosR commented on docs/uglify-options.md:
```diff
@@ -139,3 +139,13 @@ Type: `Boolean`
 Pass this flag if you don't care about full compliance with Internet Explorer 6-8 quirks.
+
+## quote_style
+Type: `Integer`
```
Needs two trailing spaces for a line break.

XhmikosR commented on tasks/lib/uglify.js:
```diff
@@ -203,6 +203,10 @@ exports.init = function(grunt) {
       outputOptions.ascii_only = options.ASCIIOnly;
     }
 
+    if (options.quote_style !== undefined) {
```
Maybe `_.isUndefined(options.quote_style)`?

aendrew commented:
Apols for the delay, will update the commit first thing tomorrow!

aendrew commented:
@XhmikosR PR updated. Also changed the "typeof varName !== 'undefined" stanzas starting at `tasks/lib/uglify.js` to use `_.isUndefined()` for consistency, hope that's okay.

XhmikosR commented:
@aendrew: please rebase and sort https://github.com/gruntjs/grunt-contrib-uglify/pull/306#discussion_r26437557 too.

aendrew commented:
Hey @XhmikosR,
I fixed the merge conflicts (and added two spaces) by merging master into my branch and then using git mergetool; though I'm totally lost as to how I should rebase now that there have been commits between when my initial and updated commits. Normally I just do `git rebase -i HEAD~2`, but that'll rewrite the 8 or 9 commits since then. Apologies for the utterly newbie question, but how do I rebase such that my commits can be squashed and the other history is preserved? Thanks!

XhmikosR commented:
Plain git rebase -i  should do the job. Otherwise 0ing me and I'll do it
myself.
On Apr 5, 2015 8:02 PM, "Ændrew Rininsland" <notifications@github.com>
wrote:

> Hey @XhmikosR <https://github.com/XhmikosR>,
> I fixed the merge conflicts (and added two spaces) by merging master into
> my branch and then using git mergetool; though I'm totally lost as to how I
> should rebase now that there have been commits between when my initial and
> updated commits. Normally I just do git rebase -i HEAD~2, but that'll
> rewrite the 8 or 9 commits since then. Apologies for the utterly newbie
> question, but how do I rebase such that my commits can be squashed and the
> other history is preserved? Thanks!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/gruntjs/grunt-contrib-uglify/pull/306#issuecomment-89812106>
> .
>

aendrew commented:
I think I've got it now — I merged upstream/master into local master, renamed my local working branch to "temp", checked out a new `293_adding_quote_style_option` branch from `master`, did a `git merge --squash temp` on `293...` and then committed afresh, force-pushing to my branch. 

Did I totally overthink this? Should I have just, in the beginning, pulled upstream master into my current branch, merged, resolved merge conflict, committed, and then squashed my merge commit into the one before it?

Thanks for taking the time to explain this to me — I should really know this, given how many open source projects I contribute to... :sweat_smile: Thanks!

XhmikosR commented on docs/uglify-options.md:
```diff
@@ -167,3 +167,14 @@ A string that is a path to a JSON cache file that uglify will create and use to
 
+
+## quote_style
+Type: `Integer`
```
Needs 2 trailing spaces for a line break.

XhmikosR commented:
It looks fine now. The only downside the way you did it is that the commit date changed, but it's not important.

aendrew commented:
@XhmikosR I think I've finally fixed the whitespace issue, now, finally. :sweat_smile:

XhmikosR commented:
Nope :tongue:

XhmikosR commented on Gruntfile.js:
```diff
@@ -326,6 +326,30 @@ module.exports = function(grunt) {
         }
+      },
+      quotes_single: {
+          files: {
```
Oh, also, please use 2 spaces for indentation consistently.

XhmikosR commented on README.md:
```diff
@@ -188,7 +188,18 @@ Type: `Boolean`
 Pass this flag if you don't care about full compliance with Internet Explorer 6-8 quirks.
-
+  
+#### quote_style
```
Finally, `quoteStyle` seems consistent with the scheme that's already being used, i.e. camelcase.

XhmikosR commented:
@aendrew: those last comments should be it. Thanks for the PR :+1:

aendrew commented:
@XhmikosR Is the whitespace in the rest of README.md fine? Because I have invisibles on and the only lines with two spaces and a newline are my lines.

To confirm I'm understanding correctly — on lines that are only newline, you want `\s\s\n`, not simply `\n`?

Updating rest of PR now.

XhmikosR commented:
In markdown two trailing spaces means a line break, so yeah.
On Apr 5, 2015 11:08 PM, "Ændrew Rininsland" <notifications@github.com>
wrote:

> @XhmikosR <https://github.com/XhmikosR> Is the whitespace in the rest of
> README.md fine? Because I have invisibles on and the only lines with two
> spaces and a newline are my lines.
>
> To confirm I'm understanding correctly — on lines that are only newline,
> you want \s\s\n, not simply \n?
>
> Updating rest of PR now.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/gruntjs/grunt-contrib-uglify/pull/306#issuecomment-89846111>
> .
>

aendrew commented:
Yay! Thanks! :+1: :shipit:

XhmikosR commented:
I added the line breaks myself in https://github.com/gruntjs/grunt-contrib-uglify/commit/672d94aae0097b8569c243a5a5164bb8f01c4bc7

aendrew commented:
@XhmikosR Thanks for doing that, not sure why I was having such issues — even after disabling the Atom Whitespace plugin.

XhmikosR commented:
New stuff is nice and all, but I'm old school and I use [my Notepad2 fork](https://github.com/XhmikosR/notepad2-mod)  :)
